catalog/core/catalog-core-solr/src/main/java/ddf/catalog/source/solr/SolrMetacardClientImpl.java

```diff
@@ -311,6 +340,74 @@ private Boolean userSpellcheckIsOn(QueryRequest request) {
     return userSpellcheckChoice;
   }
 
+  private Boolean userHighlightIsOn(QueryRequest request) {
+    Boolean userHighlight = false;
+    if (request.getProperties().get(HIGHLIGHT_KEY) != null) {
+      userHighlight = (Boolean) request.getProperties().get(HIGHLIGHT_KEY);
+    }
+    return userHighlight;
+  }
+
+  private void enableHighlighter(SolrQuery query) {
+    query.setParam(SOLR_HIGHLIGHT_KEY, true);
+    query.setParam("hl.fl", "*");
+    query.setParam("hl.simple.pre", "<b>");
+    query.setParam("hl.tag.pre", "<b>");
+    query.setParam("hl.simple.post", "</b>");
+    query.setParam("hl.tag.post", "</b>");
+    query.setParam("hl.requireFieldMatch", true);
+    query.setParam("hl.method", "unified");
+  }
+
+  private void extractHighlighing(QueryResponse response, Map<String, Serializable> responseProps) {
+    Map<String, Map<String, List<String>>> highlights = response.getHighlighting();
+    if (highlights == null) {
+      return;
+    }
+    Map<String, Map<String, Set<String>>> resultsHighlights = new HashMap<>();
+    for (String metacardId : highlights.keySet()) {
+      Map<String, List<String>> fieldHighlight = highlights.get(metacardId);
+      Map<String, Set<String>> consolidated = new HashMap<>();
+      if (fieldHighlight != null && fieldHighlight.size() > 0) {
+        for (Map.Entry<String, List<String>> entry : fieldHighlight.entrySet()) {
+          String solrField = entry.getKey();
+          String normalizedKey = resolver.resolveFieldName(solrField);
+          if (isHighlightBlacklisted(normalizedKey) || isHighlightBlacklisted(solrField)) {
+            continue;
+          }
+          Set<String> consolidatedSet = consolidated.get(normalizedKey);
+          if (consolidatedSet == null) {
+            consolidatedSet = new HashSet<>();
+          }
+          consolidatedSet.addAll(entry.getValue());
+          if (!consolidatedSet.isEmpty()) {
+            consolidated.put(normalizedKey, consolidatedSet);
+          }
+        }
+      }
+      if (!consolidated.isEmpty()) {
+        resultsHighlights.put(metacardId, consolidated);
+      }
+    }
+
+    if (!resultsHighlights.isEmpty()) {
+      try {
+        responseProps.put(HIGHLIGHT_KEY, mapper.writeValueAsString(resultsHighlights));
```

Comment:
:question: Why do you convert to JSON here instead of keeping it a Java object?  If we leave it as a Java object, different transformers and endpoints avoid converting JSON back to Java if they want to serialize to a different format or to a different JSON schema.

See examples for other response properties from the Solr provider.  I think this response should be consistent with those response properties.
https://github.com/codice/ddf/blob/73312d1/catalog/core/catalog-core-solr/src/main/java/ddf/catalog/source/solr/SolrMetacardClientImpl.java#L283-L294

Reply:
This was a shortcut on my part since the data is a map of maps and not inherently serializable.  I can fix it to send back a custom Java object.

Reply:
I created new interfaces and impl objects that get passed back in the response properties rather than converting to JSON.  I used the facet query stuff for guidelines on where to add these interfaces and classes.